torch_xla/core/xla_model.py

```diff
@@ -691,17 +692,34 @@ def add_step_closure(closure, args=()):
   Args:
     closure (callable): The function to be called.
     args (tuple): The arguments to be passed to the closure.
+    run_async: If True, run the closure asynchronously
   """
   devctx = _get_device_context()
-  step_closures = getattr(devctx, 'step_closures', None)
-  if step_closures is None:
-    step_closures = []
-    devctx.step_closures = step_closures
-  step_closures.append(lambda a=args: closure(*a))
+  if not run_async:
+    step_closures = getattr(devctx, 'step_closures', None)
+    if step_closures is None:
+      step_closures = []
+      devctx.step_closures = step_closures
+    step_closures.append(lambda a=args: closure(*a))
+  else:
+    async_step_closures = getattr(devctx, 'async_step_closures', None)
+    if async_step_closures is None:
+      async_step_closures = []
+      devctx.async_step_closures = async_step_closures
+    async_step_closures.append(lambda a=args: closure(*a))
 
 
 def _run_step_closures():
   devctx = _get_device_context()
+  async_step_closures = getattr(devctx, 'async_step_closures', None)
+  if async_step_closures is not None:
+    devctx.async_step_closures = []
+    async_closure_handler = getattr(devctx, 'async_closure_handler', None)
+    if async_closure_handler is None:
+      async_closure_handler = AsyncClosureHandler()
```

Comment:
```suggestion
      async_closure_handler = xu.closures.AsyncClosureHandler()
```
then you don't need the other import above, since xu is already imported

Reply:
fixed